### ISSUE TITLE
Suggested sampling approach for a fixed anchor

### ISSUE THREAD

**Issue by preachermaan** [KevinMusgrave/pytorch-metric-learning#31]
Hi there, I'm wondering if you have any examples of sampling where the anchor image is always fixed?

Specifically I'm looking at product recognition and in the production scenario I will always have a high quality stock image, and the similarity search will be done against lower quality phone images. I have a dataset of approx 1000 stock images and for each of these I have 5-10 lower quality images, so intuitively it feels like the anchor image should always be the stock image as this will be the case in production.

I guess the obvious choice is to do this programatically with offline triplet mining, but if there's anything available already then this will be a great help :-) 

Thanks in advance.

**Comment by KevinMusgrave** [comment]
I think this should work
```python
from pytorch_metric_learning import losses, miners

miner_func = miners.TripletMarginMiner(margin=0.1)
loss_func = losses.TripletMarginLoss(margin=0.1)

# in your training loop:
stock_batch_size = stock_imgs.size(0)
stock_emb = net(stock_imgs)
phone_emb = net(phone_imgs)

# This gets triplets where all anchors are from stock_emb
# The positives and negatives will be from phone_emb  
triplets = miner_func(stock_emb, stock_labels, phone_emb, phone_labels)

all_emb = torch.cat([stock_emb, phone_emb], dim=0)
all_labels = torch.cat([stock_labels, phone_labels], dim=0)

# You have to shift the indices for the positives and negatives to match their location in all_emb
triplets = (triplets[0],) + tuple([x+stock_batch_size if len(x) > 0 else x for x in triplets[1:]])
loss = loss_func(all_emb, all_labels, triplets)
```

**Comment by preachermaan** [comment]
Thanks @KevinMusgrave, appreciate the help and the quick response. I think I follow - I'll get everything setup and let you know how I get on. 

By the way, great library and paper. Was quite taken back by your normalised version of progress i.e. figure 4 showing reported progress vs reality ☺️🤔

**Comment by KevinMusgrave** [comment]
@preachermaan Thanks! I'll close this for now. You can comment here and/or re-open if you need help.

**Comment by marijnl** [comment]
I faced the exact same challenge so I tried to implement the suggested by Kevin, and pushed changes to my forked repo https://github.com/marijnl/pytorch-metric-learning

I aimed to stay as close to the metric learning architecture as possible. But im not fully confident I covered multiple scenarios.

**Comment by KevinMusgrave** [comment]
@marijnl If your custom trainer and tester seem to work, feel free to make a pull request, and I can take a closer look when I get time. I think this could be a useful addition to the library.

**Comment by marijnl** [comment]
Let me test a bit more and ill create a PR. 
@KevinMusgrave do you have a suggestion on how to name this update?
Now i quickly named it TwoStream but there might be some terminology from literature im not aware of.
